File: packages/python/qsu/array/_comparableKey.py
import json
import math


def _comparableKey(value) -> str:
	"""(Private) A stable string key for a value, so that array membership can be decided by
	value instead of by identity. Python cannot hash a list or a dict at all, so an identity
	comparison could not have been ported to all three languages.
	"""
	if value is None:
		return 'null'

	# `bool` is a subclass of `int`, so it has to be answered before the numeric branch.
	if isinstance(value, bool):
		return 'boolean:true' if value else 'boolean:false'

	if isinstance(value, (int, float)):
		if isinstance(value, float):
			if math.isnan(value):
				return 'number:NaN'
			if math.isinf(value):
				return 'number:Infinity' if value > 0 else 'number:-Infinity'
			# JavaScript has no int/float distinction, so `1` and `1.0` are one value there.
			if value == int(value) and abs(value) < 1e18:
				return f'number:{int(value)}'

		return f'number:{value}'

	if isinstance(value, str):
		return f'string:{value}'

	try:
		return 'json:' + json.dumps(value, default=str, separators=(',', ':'))
	except (TypeError, ValueError):
		# A cyclic structure has no JSON form.
		return f'json:{value}'

File: packages/javascript/lib/array/_comparableKey.ts
/**
 * (Private, not exported from the package) A stable string key for a value, so that array
 * membership can be decided by value instead of by reference.
 *
 * Python cannot hash a list or a dict at all, so a reference comparison could not have been
 * ported to all three languages. Comparing by value also means `arrDifference([[1]], [[1]])`
 * answers the same way everywhere.
 */
export function comparableKey(value: any): string {
	if (value === null) {
		return 'null';
	}

	if (value === undefined) {
		return 'undefined';
	}

	const type = typeof value;

	if (type === 'number') {
		if (Number.isNaN(value)) {
			return 'number:NaN';
		}

		// `-0` and `0` are one value, and `NaN` matches `NaN`, exactly like `Set` membership.
		return `number:${value === 0 ? 0 : value}`;
	}

	if (type === 'string' || type === 'boolean' || type === 'bigint') {
		return `${type}:${value}`;
	}

	if (type === 'function' || type === 'symbol') {
		// Neither has a JSON form, so they would otherwise all collapse onto one key.
		return `${type}:${String(value)}`;
	}

	try {
		return `json:${JSON.stringify(value)}`;
	} catch {
		// A cyclic structure has no JSON form.
		return `json:${String(value)}`;
	}
}
